src/email/outbound.js

```diff
@@ -0,0 +1,326 @@
+        for (let i = 0; i < addressesLength; i++) {
+            results.push(ParseEmailAddress.parseOneAddress(addresses[i]).address);
+        }
+        return results;
+    }
+
+    /**
+     * Resolve MX records by domain.
+     * @param domain
+     * @returns {Promise<MxRecord[]|*[]>}
+     */
+    async function resolveMX(domain) {
+        let resolvedMX = [];
+        if (smtpHost !== '') {
+            resolvedMX.push({ exchange: smtpHost });
+            return resolvedMX;
+        }
+        try {
+            resolvedMX = await resolver.resolveMx(domain);
+            resolvedMX.sort(function (a, b) { return a.priority - b.priority; });
+            return resolvedMX;
+        }
+        catch (ex) {
+            throw Error(`Failed to resolve MX for ${domain}: ${ex}`);
+        }
+    }
+
+    async function sendToSMTP(domain, srcHost, from, recipients, body) {
+        const resolvedMX = await resolveMX(domain);
+        log.info("Resolved mx list:", resolvedMX);
+        let sock;
+        function tryConnect(i) {
+            if (i >= resolvedMX.length) {
+                throw Error(`Could not connect to any SMTP server for ${domain}`);
+            }
+
+            sock = createConnection(smtpPort, resolvedMX[i].exchange);
+
+            sock.on('error', function (err) {
+                log.error('Error on connectMx for: ', resolvedMX[i], err);
+                tryConnect(++i);
+            });
+
+            sock.on('connect', function () {
+                log.debug('MX connection created: ', resolvedMX[i].exchange);
+                sock.removeAllListeners('error');
+                return sock;
+            });
+        }
+        tryConnect(0);
+
+        function w(s) {
+            log.debug(`SEND ${domain}> ${s}`);
+            sock.write(s + CRLF);
+        }
+
+        sock.setEncoding('utf8');
+
+        sock.on('data', function (chunk) {
+            data += chunk;
+            parts = data.split(CRLF);
+            const partsLength = parts.length - 1;
+            for (let i = 0, len = partsLength; i < len; i++) {
+                onLine(parts[i]);
+            }
+            data = parts[parts.length - 1];
+        });
+
+        sock.on('error', function (err) {
+            throw Error(`Failed to connect to ${domain}: ${err}`);
+        });
+
+        let data = '';
+        let step = 0;
+        let loginStep = 0;
+        const queue = [];
+        // const login = [];
+        let parts;
+        let cmd;
+        let upgraded = false;
+        let isUpgradeInProgress = false;
+
+        queue.push('MAIL FROM:<' + from + '>');
+        const recipientsLength = recipients.length;
+        for (let i = 0; i < recipientsLength; i++) {
+            queue.push('RCPT TO:<' + recipients[i] + '>');
+        }
+        queue.push('DATA');
+        queue.push('QUIT');
+        queue.push('');
+
+        function response(code, msg) {
+            switch (code) {
+                case 220:
```

Comment:
Having the comments next to these is great, as it means I can tell what these codes do. I would go even further and maintain a const above of `smtpCodes`:

```
const smtpCodes = {
    ServerReady: 220,
    Bye: 221,
    ...
};

and then in your code you can just do smtpCodes.ServerReady (which means people like me who don't know SMTP can read it better :D)